### ISSUE TITLE
Health API reports incorrect user actions

### ISSUE THREAD

**Issue by andreidan** [elastic/elasticsearch#86478]
### Elasticsearch Version

master

### Java Version

_bundled_

### OS Version

Darwin

### Problem Description

The health API does not distinguish between "more nodes are needed in a tier" and "an index has conflicting tiers and node attribute routing configuration, AND its target tier is missing".

For the latter, the first step is to migrate to data tiers and then enable the tier its targeting. So I believe it should report 2 user actions for the same index.

### Steps to Reproduce

On a 1 instance Elasticsearch with roles [master, data_content, data_hot] create the following indices:
```
PUT datatiers_problem_index
{
    "settings": {
        "index.routing.allocation.include._tier_preference": "data_warm",
        "index.routing.allocation.require.data": "hot"

    }
}

PUT warm-index-000001
{
  "settings": {
    "index": {
      "number_of_shards": 3,  
      "routing.allocation.include._tier_preference": "data_warm"
    }
  }
}
```

Drill down into the shards availability component:
`GET _internal/_health/data/shards_availability`

returns (notice only one user action for index `datatiers_problem_index`

```
{
  "cluster_name": "elasticsearch",
  "components": {
    "data": {
      "indicators": {
        "shards_availability": {
          "status": "red",
          "summary": "This cluster has 4  unavailable primaries, 9 unavailable replicas.",
          "details": {
            "initializing_primaries": 0,
            "restarting_replicas": 0,
            "restarting_primaries": 0,
            "unassigned_replicas": 9,
            "creating_primaries": 0,
            "started_primaries": 6,
            "unassigned_primaries": 4,
            "initializing_replicas": 0,
            "started_replicas": 0
          },
          "impacts": [
            {
              "severity": 1,
              "description": "Cannot add data to 2 indices [datatiers_problem_index, warm-index-000001]. Searches might return incomplete results.",
              "impact_areas": [
                "ingest",
                "search"
              ]
            },
            {
              "severity": 2,
              "description": "Searches might return slower than usual. Fewer redundant copies of the data exist on 3 indices [another_one, another_one2, my-index-000001].",
              "impact_areas": [
                "search"
              ]
            }
          ],
          "user_actions": [
            {
              "message": "Elasticsearch isn't allowed to allocate shards from these indices because the indices expect to be allocated to data tier nodes, but there were not any nodes with the expected tiers found in the cluster. Add nodes with the [data_warm] role to the cluster.",
              "affected_resources": [
                "datatiers_problem_index",
                "warm-index-000001"
              ]
            }
          ]
        }
      }
    }
  }
}

```

**Comment by jbaiera** [comment]
I don't know for sure if this is actually a bug or not - The shards are unassigned for one reason only and that is because they require only a warm node. That warm node doesn't exist, so there is no way for us to make an assertion that it _necessarily_ will violate the filter requirement. I suppose it would be strange to require the data attribute to be "hot" while only allowing the node type to be "warm" but should we be trying to make that prediction in the health api? It seems as if we're straying out of "Why is this shard currently unassigned" and into "Your configuration is not recognized as a best practice and might end up with a shard unassigned, you should probably change it".

**Comment by andreidan** [comment]
@jbaiera That's a good point. However, the currently indicated solution will not fix the problem for `datatiers_problem_index` and I think that's ultimately the issue we're trying to fix here.

I think we can look at the problems independently (hence why I thought we should report them both)
1) mixing the require.data and _tier_preference - when encountered we should just recommend migrating to tiers 
2) looking at _tier_preference to identify the target tier and noticing the tier is missing completely (ie. ACTION_ENABLE_TIERS_LOOKUP )

**Comment by jbaiera** [comment]
> However, the currently indicated solution will not fix the problem for datatiers_problem_index and I think that's ultimately the issue we're trying to fix here.

I agree that if they add a warm node and it does not have the correct attributes on it, that will indeed lead to another problem that needs solving - but that isn't specifically the root cause of the shard being unassigned _at this moment_. This would be a user action that is trying to prevent a problem that does not _necessarily_ exist yet. A highly likely problem, yes, but still a prediction. I think we should be wary of these kinds of additions.

An example of how I see this getting out of hand: Assume the same case, but instead of attributes, the shard limits are set too low such that adding one warm node is not going to be enough to allocate all primaries for an index on a tier. Should we add a user action to have them preemptively adjust those shard limits? Should there be another user action that says "Hey, don't just add 1 warm node. With your current settings you should add 3 or lower this setting"?

Granted, this is a slippery slope fallacy - In our cloud deployment, adding a warm node will almost certainly result in this shard being unallocated still, but I do want to highlight where we might be getting too in the weeds. This kind of comprehensiveness is easy to look at as a goal (Wow, thanks health api, I almost only added 1 node!), but it does become increasingly hard to account for all these cases in a graceful way the more comprehensive we try to be.

**Comment by jbaiera** [comment]
I'm happy to start working on a fix for this, but I think we need figure out some things:

1. Should we return user actions for indices that have conflicting tier settings, or for _any_ index that can be migrated?
2. Do we surface user actions only on indices that have shards unassigned, or for _any_ index meeting the previous requirement (GREEN or not).

I presume that the answers to the above questions are that we will want to check all indices for this, regardless of shards assigned and regardless of currently conflicting, because at any time ILM might come along and throw the system into a bad state.

Which brings me to my final question: If the idea is that we _do_ want to test indices for these combinations of settings, does it make sense to do this in a new indicator? There's a few strong supporting arguments for it in my opinion:

1. The shards availability indicator should be JUST for reporting on shards being currently available, not on warning about allocation settings that might not be safe (Single responsibility).

4. Checking for this situation does not require shard assignment information, nor does it require the explaining of allocation deciders. All we need is the index settings.

5. Indicators are pluggable, and a new indicator can live in the same module as the migrate data tiers action, thus allowing us to reuse code between the two (DRY).

6. A new indicator allows us to return a `YELLOW` or `RED` response that is orthogonal to the shard availability one. In the new indicator, a `YELLOW` could report on indices that should be migrated, but won't have issues allocating right now. A `RED` response could be used to indicate that the tier preference and data attribute filter currently conflict, regardless of if shards are still assigned (they're likely on the wrong nodes).

What do you think @andreidan ?

**Comment by andreidan** [comment]
Thanks for your thoughts @jbaiera 

I don't think we should investigate assigned shards, but you bring up some good points. 

I initially thought this falls into the "data tiers problems" category (which is something we have already). However, why stop at routing rules conflict issue and not go further to `total_shards_per_node` in combination with a missing tier - which I don't think we should do :)
